File: pacing/calculator/dtos/sub_split_mode.py
from dataclasses import dataclass


@dataclass
class SubSplitMode:
    def sub_splits(self, distance: float, moving_speed: float | None = None, pace: float | None = None) -> list[float]:
        raise NotImplementedError('sub_splits method must be implemented by subclasses')


@dataclass
class EvenSubSplitMode(SubSplitMode):
    sub_split_count: int

    def sub_splits(self, distance, moving_speed=None, pace=None) -> list[float]:
        return [distance / self.sub_split_count for _ in range(self.sub_split_count)]


@dataclass
class FixedDistanceSubSplitMode(SubSplitMode):
    sub_split_distance: float
    last_sub_split_threshold: float | None = None

    def sub_splits(self, distance, moving_speed=None, pace=None) -> list[float]:
        full_sub_split_count = int(distance // self.sub_split_distance)
        if full_sub_split_count == 0:
            return [distance]

        splits = [self.sub_split_distance for _ in range(full_sub_split_count)]

        residual_distance = distance % self.sub_split_distance
        # residual distance is within threshold, add it to last split
        if self.last_sub_split_threshold is not None and 0 < residual_distance < self.last_sub_split_threshold:
            splits[-1] = self.sub_split_distance + residual_distance
        elif 0 < residual_distance:
            splits.append(residual_distance)

        return splits


@dataclass
class CustomSubSplitMode(SubSplitMode):
    """
    A SubSplitMode that allows for custom-defined sub-split distances.
    NOTE: The sum of sub_split_distances should equal to or greater than the total distance of the split.
    If that is not the case, the remaining distance will be treated as a final sub-split.
    If the sum exceeds the total distance, the extra distances will yield invalid calculations.
    """
    sub_split_distances: list[float]

    def sub_splits(self, distance, moving_speed=None, pace=None) -> list[float]:
        return self.sub_split_distances


@dataclass
class HourSubSplitMode(SubSplitMode):
    """
    A SubSplitMode where each sub-split represents one elapsed hour (moving + down time).
    The distance per sub-split equals the pace (distance/elapsed-hour), which accounts
    for down time within the split. Falls back to moving_speed when pace is not provided.
    The final sub-split covers whatever distance remains.
    """

    def sub_splits(self, distance, moving_speed=None, pace=None) -> list[float]:
        if moving_speed is None or moving_speed <= 0:
            return [distance]

        # Use pace (dist/elapsed-hour) so boundaries fall on wall-clock hour marks.
        dist_per_hour = pace or moving_speed
        full_count = int(distance // dist_per_hour)

        if full_count == 0:
            return [distance]

        splits = [dist_per_hour for _ in range(full_count)]
        residual = distance - (full_count * dist_per_hour)
        if residual > 1e-9:
            splits.append(residual)

        return splits

File: frontend/src/calculator/subSplitMode.ts
import type { SplitPayload } from "../types";

/**
 * Given a split's sub-split configuration and its (per-split, normalized) distance,
 * returns an array of individual sub-split distances.
 *
 * @param downTimeRatio - Effective down-time ratio for the split (down_time / moving_time).
 *   Used by the "hour" mode: distPerHour = movingSpeed / (1 + downTimeRatio), so each
 *   sub-split boundary falls exactly one elapsed hour (moving + down time) apart.
 *   When omitted, defaults to 0 (no down time).
 */
export function computeSubSplitDistances(
  split: SplitPayload,
  movingSpeed?: number,
  downTimeRatio?: number,
): number[] {
  const { distance } = split;

  switch (split.sub_split_mode) {
    case "even": {
      const count = split.sub_split_count ?? 1;
      return Array.from({ length: count }, () => distance / count);
    }

    case "fixed": {
      const fixedDist = split.sub_split_distance ?? distance;
      const threshold = split.last_sub_split_threshold ?? null;
      const fullCount = Math.floor(distance / fixedDist);

      if (fullCount === 0) return [distance];

      const splits = Array.from({ length: fullCount }, () => fixedDist);
      const residual = distance % fixedDist;

      if (threshold != null && residual > 0 && residual < threshold) {
        // Absorb small residual into the last full sub-split
        splits[splits.length - 1] = fixedDist + residual;
      } else if (residual > 0) {
        splits.push(residual);
      }

      return splits;
    }

    case "custom":
      return split.sub_split_distances ?? [distance];

    case "hour": {
      if (!movingSpeed || movingSpeed <= 0) return [distance];

      // distPerHour = movingSpeed / (1 + dtr) so each sub-split spans exactly
      // one elapsed hour: moving time 1/(1+dtr) h + down time dtr/(1+dtr) h = 1 h.
      const distPerHour = movingSpeed / (1 + (downTimeRatio ?? 0));
      const fullCount = Math.floor(distance / distPerHour);

      if (fullCount === 0) return [distance];

      const splits = Array.from({ length: fullCount }, () => distPerHour);
      const residual = distance - fullCount * distPerHour;
      if (residual > 1e-9) {
        splits.push(residual);
      }

      return splits;
    }
  }
}
